src/TypeDefinition.js

```diff
@@ -54,42 +56,68 @@ export type NavigationRoute = {
   params?: NavigationParams,
 };
 
+export type NavigationStateRoute = {
+  /**
+   * React's key used by some navigators. No need to specify these manually,
+   * they will be defined by the router.
+   */
+  key: string,
+  /**
+   * For example 'Home'.
+   * This is used as a key in a route config when creating a navigator.
+   */
+  routeName: string,
+  /**
+   * Path is an advanced feature used for deep linking and on the web.
+   */
+  path?: string,
+  /**
+   * Params passed to this route when navigating to it,
+   * e.g. `{ car_id: 123 }` in a route that displays a car.
+   */
+  params?: NavigationParams,
 
-export type NavigationRouter = {
+
+  index: number,
+  routes: Array<NavigationRoute>,
+};
+
+
+export type NavigationRouter<State, Action, Options> = {
```

Comment:
This extensible router type seems much closer to the real thing! I'm glad to see that the options are a part of the router definition.

One concern: NavigationState is now customizable, but NavigationRoute is still used inside `getScreenOptions`. So I don't think this type is perfect yet.